src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -0,0 +1,390 @@
+        logging.info("Successfully inserted %d blocks into Rucio", newBlocks)
+        return newBlocks
+
+    # TODO: this will likely go away once the phedex to rucio migration is over
+    def _isBlockTierAllowed(self, blockName):
+        """
+        Check the block datatier against the list of data tiers that we want
+        to get pinned in Rucio through a block rule
+        :return: boolean whether we can or not insert a rule
+        """
+        endBlock = blockName.rsplit('/', 1)[1]
+        endTier = endBlock.split('#')[0]
+        if endTier in self.skipRulesForTiers:
+            return False
+        return True
+
+    def insertBlockRules(self, blocksAndRSEs):
+        """
+        Creates a simple replication rule for every single block that
+        has been previously inserted into Rucio.
+        Right now, we make standard rules against a given site.
+        :param blocksAndRSEs: dictionary with blocks as key, and RSEs as value
+        :return: a set of blocks with successfully created rules
+        """
+        newBlocks = set()
+        if not self.createBlockRules:
+            return newBlocks
+
+        # FIXME: this could have been done in bulk, by RSE
+        logging.info("Preparing to create block rules into Rucio...")
+        for block, rse in blocksAndRSEs.items():
+            if not self._isBlockTierAllowed(block):
+                logging.info("Component configured to skip block rule for: %s", block)
+                continue
+            rseName = rse + self.testRSEs
+            # DATASET = replicates all files in the same block to the same RSE
+            resp = self.rucio.createReplicationRule(block, rseExpression="rse=%s" % rseName,
+                                                    account=self.rucioAcct, grouping="DATASET",
+                                                    comment="WMAgent production site",
+                                                    meta=self.metaData)
+            if resp:
+                logging.info("Block rule created for %s under rule id: %s", block, resp)
+                newBlocks.add(block)
+            else:
+                logging.error("Failed to create rule for block: %s", block)
+        logging.info("Successfully inserted %d rules for blocks into Rucio", newBlocks)
+        return newBlocks
+
+    def insertReplicas(self, uninjectedData):
+        """
+        Inserts replicas into Rucio and attach them to its specific block.
+        If the insertion succeeds, also switch their database state to injected.
+
+        :param uninjectedData: dictionary with blocks as key, and RSEs as value
+        """
+        # FIXME: I think we need a different data struct from the database
+        # this method is very expensive O(n^4)
+        logging.info("Preparing to insert replicas into Rucio...")
+
+        for location in uninjectedData.keys():
+            rseName = location + self.testRSEs
```

Comment:
I wonder, does the actual job wrapper have a mechanism to optionally do stageout to the test RSE? (it is a different LFN-to-PFN algo) Otherwise this function is lying to rucio about physical location of replicas.

Reply:
Yes, all the tests I have ran over the last 72h were a lie :P
As mentioned in another comment, I'm removing this logic.

Reply:
Actually I think it would be good to expand this and add the test site stageout function to the job wrapper. Then we could have a complete test system including all sites!

Reply:
Ok, keeping the functionality doesn't hurt. You might want to open a new GH issue to expand on this test site stageout though.